src/v/ssx/when_all.h

```diff
@@ -44,40 +45,52 @@ concept emplace_backable = requires(Container c) {
 template<typename ResolvedContainer, typename FutureRange>
 requires seastar::Future<typename FutureRange::value_type>
          && std::ranges::input_range<FutureRange>
-         && std::ranges::
-           output_range<FutureRange, typename FutureRange::value_type>
          && std::constructible_from<
            typename ResolvedContainer::value_type,
-           decltype(std::declval<FutureRange>().begin()->get())>
-         && detail::emplace_backable<ResolvedContainer>
-inline seastar::future<ResolvedContainer>
-when_all_succeed(FutureRange futures) {
-    for (auto& fut : futures) {
-        fut = co_await seastar::coroutine::as_future(std::move(fut));
-    }
-
+           std::remove_reference_t<
+             decltype(std::declval<FutureRange>().begin()->get())>>
+         && detail::emplace_backable<
+           ResolvedContainer,
+           std::remove_reference_t<
+             decltype(std::declval<FutureRange>().begin()->get())>>
```

Comment:
This is a bit nitpicky, but the `decltype(std::declval<>())` can be avoided, and some more range types can be satisfied by:
```suggestion
requires seastar::Future<std::ranges::range_value_t<FutureRange>>
         && std::ranges::input_range<FutureRange>
         && std::constructible_from<
           std::ranges::range_value_t<ResolvedContainer>,
           typename std::ranges::range_value_t<FutureRange>::value_type>
         && detail::emplace_backable<
           ResolvedContainer,
           typename std::ranges::range_value_t<FutureRange>::value_type>
```

Reply:
`borrowed_range` seems like the wrong tool for this job. To (over)simplify, it checks whether an object is lvalue or not.

I think that what we want is  to ensure that `FutureRange` is a [Container](https://en.cppreference.com/w/cpp/named_req/Container). Especially the part that says
> taking care of the management of the memory used by the objects it contains.

Funny enough, this is not described in technical requirements in the slightest.

To the best of my knowledge there isn't a way to identify if a pointer (or an iterator or range, by extention) is owning or not (apart from inspecting the code around it).

`ss::when_all_succeed` gets around this problem by having two (relevant) overloads
```
auto when_all_succeed(std::vector<future<T>>);
auto when_all_succeed(Iterator begin, Iterator end);
```
the `vector` version has the lifetime sorted, as `vector` is owning.
The `range` version moves the whole range into a `vector`.

For `ssx::when_all_succeed`, where the `range` version is all there is, it feels like a major pessimisation to have to move everything in a new container every time. Also it brings up the question of what container to use. Should it be hardcoded (and which) or should it be customisable.

Given the pros/cons, I think it's best to just add it as a precondition to the function, in the documentation, that the `FutureRange` needs to own the futures.

Reply:
> For ssx::when_all_succeed, where the range version is all there is, it feels like a major pessimisation to have to move everything in a new container every time. Also it brings up the question of what container to use. Should it be hardcoded (and which) or should it be customisable.
> 
> Given the pros/cons, I think it's best to just add it as a precondition to the function, in the documentation, that the FutureRange needs to own the futures.

Fair enough. I was hoping there was an easy way to have the compiler check [CP.53](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rcoro-reference-parameters) for us, I also agree the pessimization may not be the right trade-off.